From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <spottedleaf@spottedleaf.dev>
Date: Tue, 21 Apr 2020 22:01:08 -0700
Subject: [PATCH] Prevent double logins due to async logins

Since the player was not added to the player list, the
kick logic for duplicate logins would not kick in.

Fix this by using a list of pending logins.

diff --git a/src/main/java/net/minecraft/server/NetworkManager.java b/src/main/java/net/minecraft/server/NetworkManager.java
index 905d14b895a89077e92951203cca08af1790c6f1..5b468de444d4c180906e844e545e1a54948ececd 100644
--- a/src/main/java/net/minecraft/server/NetworkManager.java
+++ b/src/main/java/net/minecraft/server/NetworkManager.java
@@ -66,6 +66,20 @@ public class NetworkManager extends SimpleChannelInboundHandler<Packet<?>> {
     private static boolean enableExplicitFlush = Boolean.getBoolean("paper.explicit-flush");
     // Paper end
 
+    // Tuinity start - prevent double login due to async login
+    // This should ONLY be used to bypass the packet listener (thanks plugins) DURING LOGIN. ANY OTHER USE IS INVALID.
+    final void disconnectDuringLoginBypass(com.mojang.authlib.GameProfile profile, IChatBaseComponent message) {
+        // copied from LoginListener
+        try {
+            MinecraftServer.LOGGER.info("Disconnecting {}: {}", profile + " (" + this.getSocketAddress() + ")", message.getString());
+            this.sendPacket(new PacketPlayOutKickDisconnect(message));
+            this.close(message);
+        } catch (Exception exception) {
+            MinecraftServer.LOGGER.error("Error whilst disconnecting player", exception);
+        }
+    }
+    // Tuinity end - prevent double login due to async login
+
     public NetworkManager(EnumProtocolDirection enumprotocoldirection) {
         this.h = enumprotocoldirection;
     }
diff --git a/src/main/java/net/minecraft/server/PlayerList.java b/src/main/java/net/minecraft/server/PlayerList.java
index cf56a821cea0481d2b8312299ed80061d226f090..96a84a891ee8cdc03b0dab2e89da958330872ab7 100644
--- a/src/main/java/net/minecraft/server/PlayerList.java
+++ b/src/main/java/net/minecraft/server/PlayerList.java
@@ -93,7 +93,16 @@ public abstract class PlayerList {
         this.getIPBans().a(true);
     }
 
+    final Map<UUID, NetworkManager> pendingLogins = new java.util.HashMap<>(); // Tuinity - fix double logins due to async login
+
     public void a(NetworkManager networkmanager, EntityPlayer entityplayer) {
+        // Tuinity start - fix double logins due to async login
+        // All of the logic has to be here thanks to plugins overriding LoginListener...
+        NetworkManager duplicateLogin = this.pendingLogins.put(entityplayer.getUniqueID(), networkmanager);
+        if (duplicateLogin != null) {
+            duplicateLogin.disconnectDuringLoginBypass(entityplayer.getProfile(), new ChatMessage("multiplayer.disconnect.duplicate_login", new Object[0]));
+        }
+        // Tuinity end - fix double logins due to async login
         entityplayer.loginTime = System.currentTimeMillis(); // Paper
         GameProfile gameprofile = entityplayer.getProfile();
         UserCache usercache = this.server.getUserCache();
@@ -136,7 +145,7 @@ public abstract class PlayerList {
         // Paper start - async load spawn in chunk
         WorldServer finalWorldserver = worldserver;
         worldserver.getChunkProvider().getTickingChunkAsync(loc.getBlockX() >> 4, loc.getBlockZ() >> 4, (chunk -> { // use ticking - as it has at least 1 neighbours loaded
-            if (networkmanager.isConnected()) {
+            if (this.pendingLogins.remove(entityplayer.getUniqueID(), networkmanager) && networkmanager.isConnected()) { // Tuinity - prevent double logins due to async login
                 postChunkLoadJoin(entityplayer, finalWorldserver, nbttagcompound, networkmanager, lastKnownName, networkmanager.getSocketAddress().toString());
             }
         }));
